### ISSUE TITLE
MSVC 2015 Unable to build shared libraries

### ISSUE THREAD

**Issue by knopp** [libressl-portable/portable#211]
Project was generated with  `-G "Visual Studio 14 2015" -DBUILD_SHARED=ON ..`

```
tls.obj : error LNK2019: unresolved external symbol __imp__shutdown@8 referenced in function _tls_close
tls.obj : error LNK2019: unresolved external symbol _posix_close referenced in function _tls_close
...
```

**Comment by busterb** [comment]
Hmm, these are just references to symbols from libcrypto. Any idea how to get around that?

**Comment by knopp** [comment]
Sure. When linking the dll all dependencies need to be set. I.e. for libcrypto you need to link crypto (object files) and winsock: Also rename crypto output to crypto-static otherwise there is conflict with dll import library (crypto.lib)
```
if (BUILD_SHARED)
	add_library(crypto-objects OBJECT ${CRYPTO_SRC})
	add_library(crypto STATIC $<TARGET_OBJECTS:crypto-objects>)
	add_library(crypto-shared SHARED $<TARGET_OBJECTS:crypto-objects>)
	target_link_libraries(crypto-shared crypto Ws2_32.lib)
	set_target_properties(crypto PROPERTIES OUTPUT_NAME crypto-static)
	set_target_properties(crypto-shared PROPERTIES OUTPUT_NAME crypto)
	set_target_properties(crypto-shared PROPERTIES VERSION
		${CRYPTO_VERSION} SOVERSION ${CRYPTO_MAJOR_VERSION})
	install(TARGETS crypto crypto-shared DESTINATION lib)
else()
	add_library(crypto STATIC ${CRYPTO_SRC})
	install(TARGETS crypto DESTINATION lib)
endif()
```

I haven't looked at ssl and tls but I'd assume they will need to link crypto-shared

There is also problem with strnlen included in cmake crypto build which causes linking conflict

**Comment by busterb** [comment]
I think this is fixed in  1fde1ea a7f754a . I did not rename the .lib output, because it did not seem to help, but please let me know if I should do it a different way. Thanks!
